staging/src/k8s.io/component-base/logs/api/v1/options.go

```diff
@@ -183,6 +213,20 @@ func featureEnabled(featureGate featuregate.FeatureGate, feature featuregate.Fea
 }
 
 func apply(c *LoggingConfiguration, options *LoggingOptions, featureGate featuregate.FeatureGate) error {
+	newCounter := atomic.AddInt32(&applyCounter, 1)
+	if newCounter > 1 {
+		switch ReapplyHandling {
+		case ReapplyHandlingError:
+			return errors.New("logging configuration was already applied earlier, changing it is not allowed")
+		case ReapplyHandlingIgnore:
+			return nil
+		case ReapplyHandlingAllow:
+			// Proceed below.
+		default:
+			return fmt.Errorf("invalid value %d for ReapplyHandling", ReapplyHandling)
+		}
+	}
```

Comment:
would it make more sense to actually detect if the config being applied differed, and always error if so?

that would avoid false positives where identical config was specified, and avoid silently ignoring config changes in tests

Reply:
Some of those unit tests are elsewhere:
- [test/integration/logs/benchmark/load_test.go](https://github.com/kubernetes/kubernetes/blob/ddf958ee06e35e1e60837dfb305403d9db97aa9b/test/integration/logs/benchmark/load_test.go#L190-L199)
- [staging/src/k8s.io/component-base/logs/json/register/register_test.go](https://github.com/kubernetes/kubernetes/blob/ddf958ee06e35e1e60837dfb305403d9db97aa9b/staging/src/k8s.io/component-base/logs/json/register/register_test.go#L158)

Reply:
> Some of those unit tests are elsewhere:

I would move the unit tests, either into the package that can reset the state, or into separate integration packages, rather than enable an "ignore ineffective config" option

Reply:
> either into the package that can reset the state

That would be `k8s.io/component-base/logs/api/v1`. Then that package inherits new dependencies (like zapr) that are undesirable and/or we get circular dependencies.

> separate integration packages

Isn't that what [test/integration/logs/benchmark/load_test.go](https://github.com/kubernetes/kubernetes/blob/ddf958ee06e35e1e60837dfb305403d9db97aa9b/test/integration/logs/benchmark/load_test.go#L190-L199) is? How would such a package reset the state in `k8s.io/component-base/logs/api/v1` if we don't offer an API for that?

Hmm, perhaps we can add a `k8s.io/component-base/logs/api/v1.Reset`? That has to come with a big warning that it should only be used when no goroutines are running and is only meant for such unit tests - basically the same semantic as `ReapplyHandlingAllow`, except that it is a separate call.

Reply:
Done: https://github.com/kubernetes/kubernetes/compare/64e1273b0d3f9ef984b8022ac1c872fe4c4197d4..4c953f58ae8c66ddc8b4dc620297e45d43c284e4